pkg/apply/apply.go

```diff
@@ -0,0 +1,135 @@
+package apply
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/lyraproj/hiera/lookup"
+	"github.com/lyraproj/hiera/provider"
+	"github.com/lyraproj/lyra/cmd/lyra/ui"
+	"github.com/lyraproj/lyra/pkg/loader"
+	"github.com/lyraproj/lyra/pkg/logger"
+	"github.com/lyraproj/puppet-evaluator/eval"
+	"github.com/lyraproj/puppet-evaluator/types"
+	"github.com/lyraproj/servicesdk/serviceapi"
+	"github.com/lyraproj/servicesdk/wfapi"
+	"github.com/lyraproj/wfe/service"
+	"github.com/lyraproj/wfe/wfe"
+	enry "gopkg.in/src-d/enry.v1"
+	yaml "gopkg.in/yaml.v2"
+
+	// Ensure that lookup function properly loaded
+	_ "github.com/lyraproj/hiera/functions"
+)
+
+// Applicator is used to apply workflows
+type Applicator struct {
+	HomeDir string
+}
+
+// ApplyWorkflowWithHieraData will apply the named workflow with the supplied hiera data
+func (a *Applicator) ApplyWorkflowWithHieraData(workflowName string, hieraData map[string]string) {
+	logger := logger.Get()
+
+	// Write the hiera data to a temp file
+	f, err := ioutil.TempFile("", "data-*.yaml")
```

Comment:
I think we should avoid taking a detour with a file and YAML for the hiera data. We may just as well just give Hiera a provider function that reads the `hieraData` map directly.

Sample code:
```
func ExampleLookup_mapProvider() {
	sampleData := map[string]string {
		`a`: `value of a`,
		`b`: `value of b`}

	tp := func(ic lookup.ProviderContext, key string, _ map[string]eval.Value) (eval.Value, bool) {
		v, ok := sampleData[key]
		return types.WrapString(v), ok
	}

	lookup.DoWithParent(context.Background(), tp, nil, func(c eval.Context) {
		fmt.Println(lookup.Lookup(impl.NewInvocation(c), `a`, nil, nil))
		fmt.Println(lookup.Lookup(impl.NewInvocation(c), `b`, nil, nil))
	})

	// Output:
	// value of a
	// value of b
}
```

Reply:
Nice